### ISSUE TITLE
insights: behaviour for search queries with multiple capturing groups

### ISSUE THREAD

**Issue by leonore** [sourcegraph/sourcegraph#40844]
from https://github.com/sourcegraph/sourcegraph/pull/40714#discussion_r954887008

we need to confirm
* how an aggregation over multiple capture groups will behave (expected behaviour is it will behave like capture group insights and just append the values to the result)
* and whether this is a change worth making for the scope

/cc @joelkw @felixfbecker @vovakulikov @unclejustin

**Comment by CristinaBirkel** [comment]
Followed up with Joel on this: https://docs.google.com/document/d/1yLV7MGOEXYT3yPvBxHHm8y2-p737KIrju_QPqlhBYz4/edit?disco=AAAAe4-mLCU

Sounds like we do want to limit to the first capture group, but I agree it could be a scoping decision since it's really not clear how we'd implement this.

**Comment by felixfbecker** [comment]
As mentioned in the RFC we made the decision to only consider the first capture group (unlike the capture group insights) to keep things simpler. This was a scope decision, we can do whatever is easier. It just needs to be consistent between the aggregation and the link generation, of course.

**Comment by felixfbecker** [comment]
Could someone clarify what the currently implemented behavior for search aggregation insights is? If it is to only consider the first group and ignore all other groups, I propose we leave it as-is (and close this issue, or move it out and rephrase it to revisit alignment with capture group insights)

**Comment by leonore** [comment]
@felixfbecker at the moment it captures across all groups but doesn't separate the values.
so for example: `deploy-(\w+)-(\w+)` and `deploy-sourcegraph-managed` would tally up `sourcegraph: 1, managed: 1`
this is consistent with capture group insights.

**Comment by felixfbecker** [comment]
If this is consistent with capture group insights, then I think we should avoid any further work here to decrease scope. The original decision to only handle the first capture group was only to keep the scope small anyway, consistency is better of course.

I would say we should make a small UI change though to make the button label "Capture groups" (plural) instead of "Capture group"

Closing unless there are objections? Will update RFC too.

cc @Joelkw

**Comment by felixfbecker** [comment]
Actually, if this behavior is true, how does this work when clicking on bars? What capture group gets replaced with the literal value?

I.e. when clicking on the bar for `managed`, how does it know whether to replace the first group or the second group with `managed`?

I tested this and it seems to always replace the **first** capture group, which is then may not match anything: `deploy-(?:managed)-(\w+)` (when it would need to be `deploy-(\w+)-(?:managed)`).

So we currently have an inconsistency within the aggregation feature we need to solve, and the only way I can see that is by adjusting the way we capture to only capture the first group. I don't see a way how a bar could ever be mapped back to the right capture group if one capture group can tally up multiple bars, that seems inherently impossible.

**Comment by CristinaBirkel** [comment]
@felixfbecker that's a good callout. My first thought on how to fix this was to add a `?:` into any capture groups beyond the first one, before we find the matches. I'll try that out.

**Comment by Joelkw** [comment]
I like that thought @CristinaBirkel! I am still in favor of limiting it to the first capturing group for ease of explaining and testing the behavior, and for the usefulness reasons noted (knowing what to replace). Thanks @felixfbecker for following up here.
